modules/javafx.graphics/src/main/native-glass/gtk/glass_window_ime.cpp

```diff
@@ -23,253 +23,168 @@
-            if (xevent.keycode) {
-                //process it as a normal key
-                process_key(event);
-                break;
-            }
-            // fall-through
-        case XLookupChars:
-            buffer[len] = 0;
-            jstring str = mainEnv->NewStringUTF(buffer);
-            EXCEPTION_OCCURED(mainEnv);
-            jsize slen = mainEnv->GetStringLength(str);
-            mainEnv->CallVoidMethod(jview,
-                    jViewNotifyInputMethod,
-                    str,
-                    NULL, NULL, NULL,
-                    slen,
-                    slen,
-                    0);
-            LOG_EXCEPTION(mainEnv)
-
-            break;
+    attr = mainEnv->NewByteArray(slen);
+    CHECK_JNI_EXCEPTION(mainEnv)
+    jbyte v[slen];
+    for (int i = 0; i < slen; i++) {
+        gunichar chr = g_utf8_get_char(&preedit_text[i]);
+
+        if (g_unichar_type(chr) == G_UNICODE_MODIFIER_SYMBOL) {
+            v[i] = com_sun_glass_ui_View_IME_ATTR_CONVERTED;
+        } else {
+            v[i] = com_sun_glass_ui_View_IME_ATTR_TARGET_NOTCONVERTED;
+        }
     }
 
-    return TRUE;
-}
+    g_free(preedit_text);
 
-bool WindowContextBase::filterIME(GdkEvent * event) {
-    if (!hasIME()) {
-        return false;
-    }
+    mainEnv->SetByteArrayRegion(attr, 0, slen, v);
+    CHECK_JNI_EXCEPTION(mainEnv)
 
-    switch (event->type) {
-        case GDK_KEY_PRESS:
-        case GDK_KEY_RELEASE:
-            return im_filter_keypress(reinterpret_cast<GdkEventKey*> (event));
-        default:
-            return FALSE;
-    }
+    mainEnv->CallVoidMethod(ctx->get_jview(),
+            jViewNotifyInputMethodLinux,
+            jstr,
+            false,
+            slen,
+            0,
+            attr);
+    LOG_EXCEPTION(mainEnv)
 }
 
-//Note: this function must return int, despite the fact it doesn't conform to XIMProc type.
-// This is required in documentation of XIM
-static int im_preedit_start(XIM im_xim, XPointer client, XPointer call) {
-    (void)im_xim;
-    (void)call;
-
-    mainEnv->CallVoidMethod((jobject) client, jViewNotifyPreeditMode, JNI_TRUE);
-    CHECK_JNI_EXCEPTION_RET(mainEnv, -1);
-    return -1; // No restrictions
+void on_preedit_end(GtkIMContext *im_context, gpointer user_data) {
+    WindowContext *ctx = (WindowContext *) user_data;
+    ctx->setOnPreEdit(false);
 }
 
-static void im_preedit_done(XIM im_xim, XPointer client, XPointer call) {
-    (void)im_xim;
-    (void)call;
-
-    mainEnv->CallVoidMethod((jobject) client, jViewNotifyPreeditMode, JNI_FALSE);
-    CHECK_JNI_EXCEPTION(mainEnv);
+void on_commit(GtkIMContext *im_context, gchar* str, gpointer user_data) {
+    WindowContext *ctx = (WindowContext *) user_data;
+    ctx->commitIME(str);
 }
 
-static void im_preedit_draw(XIM im_xim, XPointer client, XPointer call) {
-    (void)im_xim;
-    (void)call;
+void WindowContextBase::commitIME(gchar *str) {
+    if (!im_ctx.on_preedit) {
+        jchar key = g_utf8_get_char(str);
```

Comment:
This block is trying to set up calls to View.notifyKey. To get the correct KeyCode requires information in the original GdkEventKey like the hardware code and modifier state (for handling Num Lock on the key pad). The simplest solution is to record a pointer to the original GdkEventKey before calling `gtk_im_context_filter_keypress` and here pass that pointer to `process_key`. It will get the KeyCode right and generate the appropriate View.notifyKey calls.

(In theory another approach is to ignore the commit and return 'false' from filterIME so the original GdkEvent isn't filtered. I tried that but it turned into a mess because I was seeing two key press events for every keystroke. `gtk_im_context_filter_keypress` filters out the original key press event without generating a commit. Instead it posts a duplicate event with a private modifier flag set and the commit signal is sent out when that second event is passed to `gtk_im_context_filter_keypress`.)

Reply:
I think the correct way is to generate a new gdk event. If I understood correctly, it's what gtk does [internally](https://gitlab.gnome.org/GNOME/gtk/-/blob/main/gtk/gtkimcontextsimple.c). I will experiment with that.

Reply:
I did change it, but probably need some more work.

I also discovered that I probably should use the `PANGO_ATTR_*` to set the `IME_ATTR_*` with boundaries.

Reply:
I did a simple logic that if filtered and commit is sent directly without a preedit it will send the event back to `process_key`. It seems to work but I'm not sure about an unknown complex scenario.